### PR TITLE
feat: grpc middleware for stats

### PR DESCRIPTION
**Pull request by achettyiitr**
# Description

- Introducing response stats for GRPC.

## Linear Ticket

- Resolves PIPE-456

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

**Review by lvrach — APPROVED** [review]

**Review by cisse21 — APPROVED** [review]

File: warehouse/api/grpc_test.go [hunk 1]
```diff
@@ -149,7 +149,7 @@ func TestGRPC(t *testing.T) {
 		triggerStore := &sync.Map{}
 		tenantManager := multitenant.New(c, mockBackendConfig)
 		bcManager := bcm.New(c, db, tenantManager, logger.NOP, stats.Default)
-		grpcServer, err := NewGRPCServer(c, logger.NOP, db, tenantManager, bcManager, triggerStore)
+		grpcServer, err := NewGRPCServer(c, logger.NOP, stats.Default, db, tenantManager, bcManager, triggerStore)
```
Comment (fracasula) [inline warehouse/api/grpc_test.go:152]: Just expressing a preference here, I think it makes more sense if we avoid these global objects in these kind of tests.
They are not needed and we can use their more isolated counterparts that are more ideal for tests (i.e. `memstats.New()`).

Comment (fracasula) [reply warehouse/api/grpc_test.go:152]: The test passes if you do:

```diff
+++ b/warehouse/api/grpc_test.go
@@ -12,8 +12,6 @@ import (
 	"testing"
 	"time"
 
-	"github.com/rudderlabs/rudder-go-kit/stats"
-
 	"github.com/rudderlabs/rudder-server/warehouse/bcm"
 
 	"github.com/golang/mock/gomock"
@@ -32,6 +30,7 @@ import (
 	"github.com/rudderlabs/rudder-go-kit/config"
 	"github.com/rudderlabs/rudder-go-kit/filemanager"
 	"github.com/rudderlabs/rudder-go-kit/logger"
+	"github.com/rudderlabs/rudder-go-kit/stats/memstats"
 	kithelper "github.com/rudderlabs/rudder-go-kit/testhelper"
 	"github.com/rudderlabs/rudder-go-kit/testhelper/docker/resource"
 	backendconfig "github.com/rudderlabs/rudder-server/backend-config"
@@ -146,10 +145,11 @@ func TestGRPC(t *testing.T) {
 			return ch
 		}).AnyTimes()
 
+		stats := memstats.New()
 		triggerStore := &sync.Map{}
 		tenantManager := multitenant.New(c, mockBackendConfig)
-		bcManager := bcm.New(c, db, tenantManager, logger.NOP, stats.Default)
-		grpcServer, err := NewGRPCServer(c, logger.NOP, stats.Default, db, tenantManager, bcManager, triggerStore)
+		bcManager := bcm.New(c, db, tenantManager, logger.NOP, stats)
+		grpcServer, err := NewGRPCServer(c, logger.NOP, memstats.New(), db, tenantManager, bcManager, triggerStore)
 		require.NoError(t, err)
 
 		tcpPort, err := kithelper.GetFreePort()
@@ -160,7 +160,7 @@ func TestGRPC(t *testing.T) {
 		listener, err := net.Listen("tcp", tcpAddress)
 		require.NoError(t, err)
 
-		server := grpc.NewServer(grpc.Creds(insecure.NewCredentials()), grpc.UnaryInterceptor(statsInterceptor(stats.Default)))
+		server := grpc.NewServer(grpc.Creds(insecure.NewCredentials()), grpc.UnaryInterceptor(statsInterceptor(stats)))
 		proto.RegisterWarehouseServer(server, grpcServer)
 
 		g, gCtx := errgroup.WithContext(ctx)
```

Comment (achettyiitr) [reply warehouse/api/grpc_test.go:152]: Good point @fracasula, I will raise a separate PR to handle it everywhere in warehouse tests to use `config.New(`) and `memstats.New()` where ever we are passing it as a dependency.

Comment (fracasula) [reply warehouse/api/grpc_test.go:152]: Feel free to just doing them whenever you come across them, it's not important :+1:

**Pull request merged**